NuRadioReco/utilities/dataservers.py

```diff
@@ -0,0 +1,115 @@
+import requests
+import os
+import filelock
+import logging
+import shutil
+from glob import glob
+
+logger = logging.getLogger('NuRadioReco.dataservers')
+
+dataservers = ["https://rnog-data.zeuthen.desy.de", "https://rno-g.uchicago.edu/data/desy-mirror"]
+
+def get_available_dataservers_by_responsetime(dataservers=dataservers):
+    """ requests a small index file from the list of dataservers and returns a list of responsive ones ordered by elapsed time """
+    response_times = []
+    available_dataservers = []
+
+    for dataserver in dataservers:
+        # get the index of the shower_library directory, because it is short
+        testdir = f"{dataserver}/shower_library/"
+        try:
+            response = requests.get(testdir, timeout=5)
+            response.raise_for_status()
+        except:
+            continue
+        response_times.append(response.elapsed)
+        available_dataservers.append(dataserver)
+    ranked_dataservers = [x for _, x in sorted(zip(response_times, available_dataservers))]
+    return ranked_dataservers
+
+def get_available_dataservers_by_timezone(dataservers=dataservers):
+    """ uses the server locations' timezones from the list of dataservers and returns the list of dataservers ordered by proximity """
+    import socket
+    import pytz
+    from datetime import datetime
+    from geolite2 import geolite2
+
+    geo = geolite2.reader()
+
+    naive = datetime.utcnow()
+    utcoffset_local = naive.astimezone().utcoffset().total_seconds()/3600
+    server_offsets = []
+    for dataserver in dataservers:
+        dataserver_ip = socket.gethostbyname(dataserver)
+        dataserver_timezone = geo.get(dataserver_ip)["location"]["time_zone"]
+        timezone = pytz.timezone(dataserver_timezone)
+        utcoffset_server = timezone.localize(naive).utcoffset().total_seconds()/3600
+
+        server_offsets.append((utcoffset_local-utcoffset_server)%12)
+
+    ranked_dataservers = [x for _, x in sorted(zip(server_offsets, dataservers))]
+    return ranked_dataservers
+
+def download_from_dataserver(remote_path, target_path, unpack_tarball=True, dataservers=dataservers, try_ordered=False):
+    """ download remote_path to target_path from the list of NuRadio dataservers """
+
+    folder = os.path.dirname(target_path)
+    if not os.path.exists(folder):
+        os.makedirs(folder)
+
+    lockfile = target_path+".lock"
+    lock = filelock.FileLock(lockfile)
+   
+    logger.warning(f"Assuring no other process is downloading. Will wait until {lockfile} is unlocked.")
+    with lock:
+        if os.path.isfile(target_path):
+            logger.warning(f"{target_path} already exists. Maybe download was already completed by another instance?")
+            return
+        elif unpack_tarball and (len(glob(os.path.dirname(target_path) + "/*.dat")) > 0): #just check if any .dat files present (similar to NuRadioProposal.py)
```

Comment:
are the data files ever updated? Or will the name get changed in that case?

Reply:
The download function just gets called if the file needs download (either because the hash sum is not the latest version or the file is not there. The `if`/`elif` here just catches the case when 1000 jobs are started and the first that aquired the lock did already perform the download...